SCons/Util.py

```diff
@@ -1468,66 +1468,155 @@ def f(self, x, y):
     setattr(obj, name, method)
 
 
-if hasattr(hashlib, 'md5'):
-    md5 = True
+# Default hash function and format. SCons-internal.
+_hash_function = None
+_hash_format = None
 
-    def MD5signature(s):
-        """
-        Generate md5 signature of a string
 
-        :param s: either string or bytes. Normally should be bytes
-        :return: String of hex digits representing the signature
-        """
-        m = hashlib.md5()
+def get_hash_format():
+    """
+    Retrieves the hash format or None if not overridden. A return value of None
+    does not guarantee that MD5 is being used; instead, it means that the
+    default precedence order documented in SCons.Util.set_hash_format is
+    respected.
+    """
+    return _hash_format
 
-        try:
-            m.update(to_bytes(s))
-        except TypeError as e:
-            m.update(to_bytes(str(s)))
 
-        return m.hexdigest()
+def set_hash_format(hash_format):
+    """
+    Sets the default hash format used by SCons. If hash_format is None or
+    an empty string, the default is determined by this function.
 
-    def MD5filesignature(fname, chunksize=65536):
-        """
-        Generate the md5 signature of a file
+    Currently the default behavior is to use the first available format of
+    the following options: MD5, SHA1, SHA256.
+    """
+    global _hash_format, _hash_function
+
+    _hash_format = hash_format
+    if hash_format:
+        hash_format_lower = hash_format.lower()
+        _hash_function = getattr(hashlib, hash_format_lower, None)
+        if _hash_function is None:
+            raise Exception(
+                'Hash format "%s" is not available in your Python '
+                'interpreter.' % hash_format_lower)
+    else:
+        # Set the default hash format based on what is available, defaulting
+        # to md5 for backwards compatibility.
+        choices = ['md5', 'sha1', 'sha256']
+        for choice in choices:
+            try:
+                _hash_function = getattr(hashlib, choice)
+                _hash_function()
+                break
+            except Exception:
+                pass
+        else:
+            # This is not expected to happen in practice.
+            raise Exception(
```

Comment:
I think we have a specific exception type for user error or command line error which would be better here.